### ISSUE TITLE
Unable to use withSortField() of SearchOperation class [JIRA: CLIENTS-627]

### ISSUE THREAD

**Issue by abdulMuneerM** [basho/riak-java-client#554]
```
searchOp = new SearchOperation
                        .Builder(BinaryValue.create(searchIndex), "*:*")
                        .withStart(0)
                        .withNumRows(10)
                        .withReturnFields("*")
                        .withSortField("fieldName_l asc")
                        .build();
cluster.execute(searchOp);
```

When i tried the above code, riak is crashing with the following crash report.
https://gist.github.com/abdulMuneerM/606c5073825018e43725

And when i changed it to "fieldName_l", getting the following 

```
2015-10-13 16:46:51.286 ERROR 11484 --- [nio-8080-exec-3] to.lookup.riak.Cluster                   : search: 

java.util.concurrent.ExecutionException: com.basho.riak.client.core.netty.RiakResponseException: Query unsuccessful check the logs.
        at com.basho.riak.client.core.FutureOperation.get(FutureOperation.java:260)
        at to.lookup.riak.Cluster.searchViaPBC(Cluster.java:173)
        at to.lookup.riak.Service.searchData(Service.java:251)
        at to.lookup.riak.Controller.searchData(Controller.java:73)
        at sun.reflect.GeneratedMethodAccessor27.invoke(Unknown Source)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:221)
        at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:137)
        at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:110)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandleMethod(RequestMappingHandlerAdapter.java:776)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:705)
        at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:85)
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:959)
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:893)
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:967)
        at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:869)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:648)
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:843)
        at javax.servlet.http.HttpServlet.service(vlet.java:729)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:52)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:77)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:85)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:502)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:518)
        at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1091)
        at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.prostractProtocol.java:668)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1521)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1478)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:745)
Caused by: com.basho.riak.client.core.netty.RiakResponseException: Query unsuccessful check the logs.
        at com.basho.riak.client.core.netty.RiakResponseHandler.channelRead(RiakResponseHandler.java:54)
        at io.netty.channel.DefaultChannelHandlerContext.invokeChannelRead(DefaultChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelHandlerContext.fireChannelRead(DefaultChannelHandlerContext.java:326)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:155)
        at io.netty.handler.codec.ByteToMessageCodec.channelRead(ByteToMessageCodec.java:108)
        at io.netty.channel.DefaultChannelHandlerContext.invokeChannelRead(DefaultChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelHandlerContext.fireChannelRead(DefaultChannelHandlerContext.java:326)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:785)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:116)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:494)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:461)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:378)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:350)
        at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:101)
        ... 1 common frames omitted
```

with _solr_ log as 

```
2015-10-13 16:46:51,283 [ERROR] <qtp101478235-17>@SolrException.java:109 org.apache.solr.common.SolrException: Can't determine a Sort Order (asc or desc) in sort spec 'fieldName_l', pos=11
    at org.apache.solr.search.QueryParsing.parseSortSpec(QueryParsing.java:339)
    at org.apache.solr.search.QParser.getSort(QParser.java:247)
    at org.apache.solr.handler.component.QueryComponent.prepare(QueryComponent.java:149)
    at org.apache.solr.handler.component.SearchHandler.handleRequestBody(SearchHandler.java:193)
    at org.apache.solr.handler.RequestHandlerBase.handleRequest(RequestHandlerBase.java:135)
    at org.apache.solr.core.SolrCore.execute(SolrCore.java:1916)
    at org.apache.solr.servlet.SolrDispatchFilter.execute(SolrDispatchFilter.java:780)
    at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:427)
    at org.apache.solr.servlet.SolrDispatchFilter.doFilter(SolrDispatchFilter.java:217)
    at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1419)
    at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:455)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:137)
    at org.eclipse.jetty.security.SecurityHandler.handle(SecurityHandler.java:557)
    at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:231)
    at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1075)
    at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:384)
    at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:193)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1009)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:135)
    at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:255)
    at org.eclipse.jetty.server.handler.HandlerCollection.handle(HandlerCollection.java:154)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)
    at org.eclipse.jetty.server.Server.handle(Server.java:368)
    at org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:489)
    at org.eclipse.jetty.server.BlockingHttpConnection.handleRequest(BlockingHttpConnection.java:53)
    at org.eclipse.jetty.server.AbstractHttpConnection.content(AbstractHttpConnection.java:953)
    at org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.content(AbstractHttpConnection.java:1014)
    at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:861)
    at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:240)
    at org.eclipse.jetty.server.BlockingHttpConnection.handle(BlockingHttpConnection.java:72)
    at org.eclipse.jetty.server.bio.SocketConnector$ConnectorEndPoint.run(SocketConnector.java:264)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:608)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:543)
    at java.lang.Thread.run(Thread.java:745)
```

How do I sort the result set using the withSortField() function ? Isn't this a bug ?

**Comment by alexmoore** [comment]
Hi @abdulMuneerM,

It should be `.withSortField("fieldName_l asc")`, and I've tested it locally so it works in my case.  

Are you using the default schema?

Thanks,
Alex

**Comment by alexmoore** [comment]
Also, what version of the Java client and what version of Riak were you using?

**Comment by empovit** [comment]
@alexmoore I can confirm that passing return fields in combination with sorting criteria causes this kind of failure. As far as I remember, it can be worked around by including "score" (Solr) in the list of return fields. Happens also when using the `new Search.Builder()` API. Java 7, Riak 2.0.x, client 2.0.x. Not tested with 2.0.2 though. Might it be a Solr bug?

@abdulMuneerM Try removing `withReturnFields()` and see if you still get the error.

**Comment by alexmoore** [comment]
Hmm, I can still get results back with `.withReturnFields("*").withSortField("fieldName_i asc")` with the latest Java client && Riak 2.1.3.  Need more info to continue :/

**Comment by empovit** [comment]
@alexmoore What info do you need? I just tested it with Riak 2.0.6 and client 2.0.2. There's absolutely no problem reproducing the issue although the error is different. 

My datatype has two fields - `number_i` and `square_i`. The results I'm getting with the default schema are as follows:

### Version 1

`SearchOperation search = new SearchOperation.Builder(BinaryValue.create(searchIndex), "*:*").withSortField("number_i asc").withStart(0).withNumRows(10).build();`

getting

`{score=[1.00000000000000000000e+00], _yz_rb=[samples_default], _yz_id=[1*default*samples_default*0*6], _yz_rk=[0], number_i=[0], square_i=[0], _yz_rt=[default]}`
`{score=[1.00000000000000000000e+00], _yz_rb=[samples_default], _yz_id=[1*default*samples_default*1*6], _yz_rk=[1], number_i=[1], square_i=[1], _yz_rt=[default]}`
`...`

### Version 2

`SearchOperation search = new SearchOperation.Builder(BinaryValue.create(searchIndex), "*:*").withReturnFields(Arrays.asList("square_i")).withSortField("number_i asc").withStart(0).withNumRows(10).build();`

getting

```
Exception in thread "main" java.util.concurrent.ExecutionException: com.basho.riak.client.core.netty.RiakResponseException: Error processing incoming message: error:badarg:[{protobuffs,encode_internal,
                                                  [2,[],float],
                                                  [{file,"src/protobuffs.erl"},
                                                   {line,167}]},
                                                 {riak_search_pb,iolist,2,
                                                  [{file,
                                                    "src/riak_search_pb.erl"},
                                                   {line,64}]},
                                                 {riak_search_pb,encode,2,
                                                  [{file,
                                                    "src/riak_search_pb.erl"},
                                                   {line,49}]},
                                                 {riak_pb_codec,encode,1,
                                                  [{file,
                                                    "src/riak_pb_codec.erl"},
                                                   {line,77}]},
                                                 {yz_pb_search,encode,1,
                                                  [{file,
                                                    "src/yz_pb_search.erl"},
                                                   {line,60}]},
                                                 {riak_api_pb_server,
                                                  send_encoded_message_or_error,
                                                  3,
                                                  [{file,
                                                    "src/riak_api_pb_server.erl"},
                                                   {line,462}]},
                                                 {riak_api_pb_server,
                                                  process_message,4,
                                                  [{file,
                                                    "src/riak_api_pb_server.erl"},
                                                   {line,394}]},
                                                 {riak_api_pb_server,
                                                  connected,2,
                                                  [{file,
                                                    "src/riak_api_pb_server.erl"},
                                                   {line,226}]}]
        at com.basho.riak.client.core.FutureOperation.get(FutureOperation.java:260)
        at riak.SearchPlayground.main(SearchPlayground.java:56)
Caused by: com.basho.riak.client.core.netty.RiakResponseException: Error processing incoming message: error:badarg:[{protobuffs,encode_internal,
                                                  [2,[],float],
                                                  [{file,"src/protobuffs.erl"},
                                                   {line,167}]},
                                                 {riak_search_pb,iolist,2,
                                                  [{file,
                                                    "src/riak_search_pb.erl"},
                                                   {line,64}]},
                                                 {riak_search_pb,encode,2,
                                                  [{file,
                                                    "src/riak_search_pb.erl"},
                                                   {line,49}]},
                                                 {riak_pb_codec,encode,1,
                                                  [{file,
                                                    "src/riak_pb_codec.erl"},
                                                   {line,77}]},
                                                 {yz_pb_search,encode,1,
                                                  [{file,
                                                    "src/yz_pb_search.erl"},
                                                   {line,60}]},
                                                 {riak_api_pb_server,
                                                  send_encoded_message_or_error,
                                                  3,
                                                  [{file,
                                                    "src/riak_api_pb_server.erl"},
                                                   {line,462}]},
                                                 {riak_api_pb_server,
                                                  process_message,4,
                                                  [{file,
                                                    "src/riak_api_pb_server.erl"},
                                                   {line,394}]},
                                                 {riak_api_pb_server,
                                                  connected,2,
                                                  [{file,
                                                    "src/riak_api_pb_server.erl"},
                                                   {line,226}]}]
        at com.basho.riak.client.core.netty.RiakResponseHandler.channelRead(RiakResponseHandler.java:54)
        at io.netty.channel.DefaultChannelHandlerContext.invokeChannelRead(DefaultChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelHandlerContext.fireChannelRead(DefaultChannelHandlerContext.java:326)
        at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:155)
        at io.netty.handler.codec.ByteToMessageCodec.channelRead(ByteToMessageCodec.java:108)
        at io.netty.channel.DefaultChannelHandlerContext.invokeChannelRead(DefaultChannelHandlerContext.java:340)
        at io.netty.channel.DefaultChannelHandlerContext.fireChannelRead(DefaultChannelHandlerContext.java:326)
        at io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:785)
        at io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:116)
        at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:494)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:461)
        at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:378)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:350)
        at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:101)
        at java.lang.Thread.run(Unknown Source)
```

and **sometimes**

```
Exception in thread "main" java.util.concurrent.ExecutionException: java.nio.channels.ClosedChannelException
    at com.basho.riak.client.core.FutureOperation.get(FutureOperation.java:260)
    at riak.SearchPlayground.main(SearchPlayground.java:56)
Caused by: java.nio.channels.ClosedChannelException
```

### Version 3

`SearchOperation search = new SearchOperation.Builder(BinaryValue.create(searchIndex), "*:*").withReturnFields(Arrays.asList("*")).withSortField("number_i asc").withStart(0).withNumRows(10).build();` 

same as **Version 2**

### Version 4

`SearchOperation search = new SearchOperation.Builder(BinaryValue.create(searchIndex), "*:*").withReturnFields(Arrays.asList("square_i", "score")).withSortField("number_i asc").withStart(0).withNumRows(10).build();`

getting

`{square_i=[0], score=[1.00000000000000000000e+00]}`
`{square_i=[1], score=[1.00000000000000000000e+00]}`
`...`

**Comment by alexmoore** [comment]
@empovit Ahh, it only shows up in Riak 2.0.x, I was using 2.1.3 - thanks for the sample code & setup.

@abdulMuneerM This is actually an issue in Riak 2.0.x: basho/yokozuna#372  A fix for it is included in Riak 2.1.3, and it should be included in Riak 2.0.7 whenever it's released.

For now you can include the "score" field in your returned fields options to get around the issue.

Thanks,
Alex
